### PR TITLE
Optimize AES-XTS mode in OpenSSL for aarch64

### PR DESCRIPTION
**Pull request by xkqian**
Aes-xts mode can be optimized by interleaving cipher operation on
several blocks and loop unrolling. Interleaving needs one ideal
unrolling factor, here we adopt the same factor with aes-cbc,
which is described as below:
	If blocks number > 5, select 5 blocks as one iteration,every
	loop, decrease the blocks number by 5.
	If 3 < left blocks < 5 select 3 blocks as one iteration, every
	loop, decrease the block number by 3.
	If left blocks < 3, treat them as tail blocks.
Detailed implementation has a little adjustment for squeezing
code space.
With this way, for small size such as 16 bytes, the performance is
similar as before, but for big size such as 16k bytes, the performance
improves a lot, even reaches to 2x uplift, for some arches such as A57,
the improvement even reaches more than 2x uplift. We collect many
performance datas on different micro-archs such as thunderx2,
ampere-emag, a72, a57, a53 and N1, all of which reach 0.5-2x uplift.
The following table lists the encryption performance data on aarch64,
take a72, a57, a53 and N1 as examples. Performance value takes the
unit of cycles per byte, takes the format as comparision of values.
List them as below:

```
A72:
                            Before optimization     After optimization  Improve
evp-aes-128-xts@16          8.899913518             5.949087263         49.60%
evp-aes-128-xts@64          4.525512668             3.389141845         33.53%
evp-aes-128-xts@256         3.502906908             1.633573479         114.43%
evp-aes-128-xts@1024        3.174210419             1.155952639         174.60%
evp-aes-128-xts@8192        3.053019303             1.028134888         196.95%
evp-aes-128-xts@16384       3.025292462             1.02021169          196.54%
evp-aes-256-xts@16          9.971105023             6.754233758         47.63%
evp-aes-256-xts@64          4.931479093             3.786527393         30.24%
evp-aes-256-xts@256         3.746788153             1.943975947         92.74%
evp-aes-256-xts@1024        3.401743802             1.477394648         130.25%
evp-aes-256-xts@8192        3.278769327             1.32950421          146.62%
evp-aes-256-xts@16384       3.27093296              1.325276257         146.81%

A57:
                            Before optimization     After optimization  Improve
evp-aes-128-xts@16          10.57455446             7.165438012         47.58%
evp-aes-128-xts@64          5.418185447             3.721241202         45.60%
evp-aes-128-xts@256         3.855184592             1.747145379         120.66%
evp-aes-128-xts@1024        3.477199757             1.253049735         177.50%
evp-aes-128-xts@8192        3.36768104              1.091943159         208.41%
evp-aes-128-xts@16384       3.360373443             1.088942789         208.59%
evp-aes-256-xts@16          12.54559459             8.745489036         43.45%
evp-aes-256-xts@64          6.542808937             4.326387568         51.23%
evp-aes-256-xts@256         4.62668822              2.119908754         118.25%
evp-aes-256-xts@1024        4.161716505             1.557335554         167.23%
evp-aes-256-xts@8192        4.032462227             1.377749511         192.68%
evp-aes-256-xts@16384       4.023293877             1.371558933         193.34%

A53:
                            Before optimization     After optimization  Improve
evp-aes-128-xts@16          18.07842135             13.96980808         29.40%
evp-aes-128-xts@64          7.933818397             6.07159276          30.70%
evp-aes-128-xts@256         5.264604704             2.611155744         101.60%
evp-aes-128-xts@1024        4.606660117             1.722713454         167.40%
evp-aes-128-xts@8192        4.405160115             1.454379201         202.90%
evp-aes-128-xts@16384       4.401592028             1.442279392         205.20%
evp-aes-256-xts@16          20.07084054             16.00803726         25.40%
evp-aes-256-xts@64          9.192647294             6.883876732         33.50%
evp-aes-256-xts@256         6.336143161             3.108140452         103.90%
evp-aes-256-xts@1024        5.62502952              2.097960651         168.10%
evp-aes-256-xts@8192        5.412085608             1.807294191         199.50%
evp-aes-256-xts@16384       5.403062591             1.790135764         201.80%

N1:
                            Before optimization     After optimization  Improve
evp-aes-128-xts@16          6.48147613              4.209415473         53.98%
evp-aes-128-xts@64          2.847744115             1.950757468         45.98%
evp-aes-128-xts@256         2.085711968             1.061903238         96.41%
evp-aes-128-xts@1024        1.842014669             0.798486302         130.69%
evp-aes-128-xts@8192        1.760449052             0.713853939         146.61%
evp-aes-128-xts@16384       1.760763546             0.707702009         148.80%
evp-aes-256-xts@16          7.264142817             5.265970454         37.94%
evp-aes-256-xts@64          3.251356212             2.41176323          34.81%
evp-aes-256-xts@256         2.380488469             1.342095742         77.37%
evp-aes-256-xts@1024        2.08853022              1.041718215         100.49%
evp-aes-256-xts@8192        2.027432668             0.944571334         114.64%
evp-aes-256-xts@16384       2.00740782              0.941991415         113.10%
```

Add more XTS test cases to cover the cipher stealing mode and cases of different
number of blocks.

### CODE REVIEW

**Comment by xkqian** [comment]
Hi, all, this optimization patch is only for aarch64(Within the patch, I add 6 test cases for xts in evpciph.txt),  but one of the travis-ci test failed on x86-64, job name:33370.5. reported error below:
```
$ if [ -z "$BUILDONLY" ]; then if [ -n "$CROSS_COMPILE" ]; then sudo dpkg --add-architecture i386; sudo apt-get update; sudo apt-get -yq --no-install-suggests --no-install-recommends --force-yes install wine; export EXE_SHELL="wine" WINEPREFIX=`pwd`; fi; if [ -e krb5/src ]; then sudo apt-get -yq install bison dejagnu gettext keyutils ldap-utils libldap2-dev libkeyutils-dev python-cjson python-paste python-pyrad slapd tcl-dev tcsh; fi; if ! HARNESS_VERBOSE_FAILURE=yes BORING_RUNNER_DIR=$top/boringssl/ssl/test/runner travis_wait 60 make _tests; then echo -e '\052\052 FAILED -- MAKE TEST'; travis_terminate 1; fi; else if ! $make build_tests >~/build.log 2>&1; then echo -e '\052\052 FAILED -- MAKE BUILD_TESTS'; cat ~/build.log travis_terminate 1; fi; fi
Still running (43 of 60): make _tests
The job exceeded the maximum time limit for jobs, and has been terminated.  
```
Config option as below:
```
AMD64
 Compiler: clang Xcode: xcode9.3 C
CONFIG_OPTS="no-asm -Werror --debug no-afalgeng no-shared enable-rc5 enable-md2 -fsanitize=address" LSAN_OPTIONS="report_objects=1"
```.
Log shows it exceeds the time limit, but the similar job 33370.6 pass the test. Can anyone tell what error happend? Or is this one known issue?

**Comment by xkqian** [comment]
> Hi, all, this optimization patch is only for aarch64(Within the patch, I add 6 test cases for xts in evpciph.txt), but one of the travis-ci test failed on x86-64, job name:33370.5. reported error as below:
> 
> ```
> $ if [ -z "$BUILDONLY" ]; then if [ -n "$CROSS_COMPILE" ]; then sudo dpkg --add-architecture i386; sudo apt-get update; sudo apt-get -yq --no-install-suggests --no-install-recommends --force-yes install wine; export EXE_SHELL="wine" WINEPREFIX=`pwd`; fi; if [ -e krb5/src ]; then sudo apt-get -yq install bison dejagnu gettext keyutils ldap-utils libldap2-dev libkeyutils-dev python-cjson python-paste python-pyrad slapd tcl-dev tcsh; fi; if ! HARNESS_VERBOSE_FAILURE=yes BORING_RUNNER_DIR=$top/boringssl/ssl/test/runner travis_wait 60 make _tests; then echo -e '\052\052 FAILED -- MAKE TEST'; travis_terminate 1; fi; else if ! $make build_tests >~/build.log 2>&1; then echo -e '\052\052 FAILED -- MAKE BUILD_TESTS'; cat ~/build.log travis_terminate 1; fi; fi
> Still running (43 of 60): make _tests
> The job exceeded the maximum time limit for jobs, and has been terminated.  
> ```
> 
> Config option as below:
> 
> ```
> AMD64
>  Compiler: clang Xcode: xcode9.3 C
> CONFIG_OPTS="no-asm -Werror --debug no-afalgeng no-shared enable-rc5 enable-md2 -fsanitize=address" LSAN_OPTIONS="report_objects=1"
> ```
> 
> Log shows it exceeds the time limit, but the similar job 33370.6 pass the test. Can anyone tell what error happend? Or is this one known issue?

From the answer from matcaswell in another PR #11405, 

```
> Does the travis-ci failure of 33398.5 affect the result? I have one new patch which has the same travis-ci failure with 33398.5. Even my patch targets on the aarch64, I am no idea why it fails on x86-64...

No. Its a timeout on that job. Its a recurring problem that we haven't figured out how to fix yet.
```
Seems it's one known issue, so we are ok to go on reviewing.

**Comment by xkqian** [comment]
Ping...  
Will anyone take a look at this aes-xts optimization?

**Comment by xkqian** [comment]
Ping again to make it rise to the surface...

File: crypto/aes/asm/aesv8-armx.pl [hunk 1]
```diff
@@ -2131,6 +2131,1460 @@ ()
...
+	aesmc	$dat0,$dat0
+	vld1.32	{q8-q9},[$key1],#32			// load key schedule...
+	aese	$dat0,q21
+	aesmc	$dat0,$dat0
+	subs	$rounds,$rounds,#10			// if rounds==10, jump to aes-128-xts processing
+	b.eq	.Lxts_128_enc
+.Lxts_enc_round_loop:
+	aese	$dat0,q8
+	aesmc	$dat0,$dat0
+	vld1.32	{q8},[$key1],#16				// load key schedule...
```
Comment (t8m) [inline crypto/aes/asm/aesv8-armx.pl:2242]: The indentation of comment is off.

File: crypto/aes/asm/aesv8-armx.pl [hunk 2]
```diff
@@ -2131,6 +2131,1460 @@ ()
...
+	extr	$ivh,$ivh,$ivl,#63
+	and	$tmpmw,$constnum,$midnum,asr#31
+	eor	$ivl,$tmpmx,$ivl,lsl#1
+	fmov	$ivd10,$ivl
+	fmov	$ivd11,$ivh
+
+	ldr	$rounds0,[$key1,#240]			// next starting point
+	vld1.8	{$dat},[$inp],$step
+
+	vld1.32	{q8-q9},[$key1]		// load key schedule...
```
Comment (t8m) [inline crypto/aes/asm/aesv8-armx.pl:2329]: Again the indentation of the comment is off.

File: crypto/aes/asm/aesv8-armx.pl [hunk 3]
```diff
@@ -2131,6 +2131,1460 @@ ()
...
+	vld1.8	{$dat2},[$inp],#16
+	subs	$len,$len,#32				// bias
+	add	$rounds,$rounds0,#2
+	vorr	$in1,$dat,$dat
+	vorr	$dat1,$dat,$dat
+	vorr	$in3,$dat,$dat
+	vorr	$in2,$dat2,$dat2
+	vorr	$in4,$dat2,$dat2
+	b.lo	.Lxts_enc_tail
+	veor	$dat,$dat,$iv0				//before encryt, xor with iv
```
Comment (t8m) [inline crypto/aes/asm/aesv8-armx.pl:2352]: Typo encryt, no space after `//`

File: crypto/aes/asm/aesv8-armx.pl [hunk 4]
```diff
@@ -2131,6 +2131,1460 @@ ()
...
+	sub	$rounds0,$rounds0,#2
+	vld1.32	{q10-q11},[$key_],#32
+	vld1.32	{q12-q13},[$key_],#32
+	vld1.32	{q14-q15},[$key_],#32
+	vld1.32	{$rndlast},[$key_]
+
+	add	$key_,$key1,#32
+	mov	$rounds,$rounds0
+
+	//encryption.
```
Comment (t8m) [inline crypto/aes/asm/aesv8-armx.pl:2341]: No space after `//`, why the dot?

File: crypto/aes/asm/aesv8-armx.pl [hunk 5]
```diff
@@ -2131,6 +2131,1460 @@ ()
...
+	aese	$dat1,q10
+	aesmc	$dat1,$dat1
+	aese	$dat2,q10
+	aesmc	$dat2,$dat2
+	aese	$dat3,q10
+	aesmc	$dat3,$dat3
+	aese	$dat4,q10
+	aesmc	$dat4,$dat4
+	add	$inp,$inp,$xoffset		// x0 is adjusted in such way that
+					// at exit from the loop v1.16b-v26.16b
```
Comment (t8m) [inline crypto/aes/asm/aesv8-armx.pl:2460]: Indentation of comment is off.

File: crypto/aes/asm/aesv8-armx.pl [hunk 6]
```diff
@@ -2131,6 +2131,1460 @@ ()
...
+	// The iv for second block
+	extr	$midnumx,$ivh,$ivh,#32
+	extr	$ivh,$ivh,$ivl,#63
+	and	$tmpmw,$constnum,$midnum,asr #31
+	eor	$ivl,$tmpmx,$ivl,lsl #1
+	fmov	$ivd10,$ivl
+	fmov	$ivd11,$ivh
+	veor	$tmp2,$iv2,$rndlast
+	add	$inp,$inp,$xoffset			// $inp is adjusted in such way that
+						// at exit from the loop $dat1-$dat2
```
Comment (t8m) [inline crypto/aes/asm/aesv8-armx.pl:2677]: Wrong indentation?

File: crypto/aes/asm/aesv8-armx.pl [hunk 7]
```diff
@@ -2131,6 +2131,1460 @@ ()
...
+	ldrb	$l2outp,[$out,$tailcnt]
+	ldrb	$loutp,[$tmpinp,$tailcnt]
+	strb	$l2outp,[$tmpoutp,$tailcnt]
+	strb	$loutp,[$out,$tailcnt]
+	b.gt	.composite_enc_loop
+.Lxts_enc_load_done:
+	vld1.8	{$tmpin},[$out]
+	veor	$tmpin,$tmpin,$iv0
+
+	// enrypt the composite block to get the last second encrypted text block
```
Comment (t8m) [inline crypto/aes/asm/aesv8-armx.pl:2830]: Typo enrypt

File: crypto/aes/asm/aesv8-armx.pl [hunk 8]
```diff
@@ -2131,6 +2131,1460 @@ ()
...
+	and	$tmpmw,$constnum,$midnum,asr #31
+	eor	$ivl,$tmpmx,$ivl,lsl #1
+	fmov	$ivd30,$ivl
+	fmov	$ivd31,$ivh
+
+	add	$key_,$key1,#32
+	mov	$rounds,$rounds0
+	b	.Lxts_dec
+
+	//decryption.
```
Comment (t8m) [inline crypto/aes/asm/aesv8-armx.pl:3051]: No space, superfluous dot.

File: crypto/aes/asm/aesv8-armx.pl [hunk 9]
```diff
@@ -2131,6 +2131,1460 @@ ()
...
+
+.Lxts_done:
+	tst	$tailcnt,#0xf
+	b.eq	.Lxts_dec_abort
+	// Processing the last two blocks with cipher stealing.
+	mov	x7,x3
+	cbnz	x2,.Lxts_dec_1st_done
+	vld1.32	{$dat0},[$inp],#16
+
+	// derypt the last secod block to get the last plain text block
```
Comment (t8m) [inline crypto/aes/asm/aesv8-armx.pl:3512]: Typo derypt

File: crypto/aes/asm/aesv8-armx.pl [hunk 10]
```diff
@@ -2131,6 +2131,1460 @@ ()
...
+	vld1.32	{$dat0},[$key1]
+	aesd	$tmpin,$dat1
+	veor	$tmpin,$tmpin,$dat0
+	veor	$tmpin,$tmpin,$iv1
+	vst1.8	{$tmpin},[$out]
+
+	mov	$tmpinp,$inp
+	add	$tmpoutp,$out,#16
+
+	// Compisite the tailcnt "16 byte not aligned block" into the last second plain blocks,
```
Comment (t8m) [inline crypto/aes/asm/aesv8-armx.pl:3540]: Typo compisite

File: crypto/aes/asm/aesv8-armx.pl (hunk 2 above)
Comment (xkqian) [reply crypto/aes/asm/aesv8-armx.pl:2329]: Push to fix this.

File: crypto/aes/asm/aesv8-armx.pl (hunk 3 above)
Comment (xkqian) [reply crypto/aes/asm/aesv8-armx.pl:2352]: Push to fix it.

File: crypto/aes/asm/aesv8-armx.pl (hunk 9 above)
Comment (xkqian) [reply crypto/aes/asm/aesv8-armx.pl]: push to fix it

File: crypto/aes/asm/aesv8-armx.pl (hunk 7 above)
Comment (xkqian) [reply crypto/aes/asm/aesv8-armx.pl]: push to fix it

File: crypto/aes/asm/aesv8-armx.pl (hunk 4 above)
Comment (xkqian) [reply crypto/aes/asm/aesv8-armx.pl]: push to fix it

File: crypto/aes/asm/aesv8-armx.pl (hunk 5 above)
Comment (xkqian) [reply crypto/aes/asm/aesv8-armx.pl]: push to fix it

File: crypto/aes/asm/aesv8-armx.pl (hunk 6 above)
Comment (xkqian) [reply crypto/aes/asm/aesv8-armx.pl:2677]: push to fix it

**Comment by xkqian** [comment]
Update patch to fix the formating and typo issues from the comment of Tomas.

**Comment by mattcaswell** [comment]
@levitte are you able to take a look at this?

**Comment by xkqian** [comment]
@levitte I am not sure who should I turn to help about the arm assembly code,  but I just remembered 
 that you have reviewd several arm patches, so I turn to you.  Can you kindly help a look at this if you are available? Thanks.

File: test/recipes/30-test_evp_data/evpciph.txt [hunk 11]
```diff
@@ -1360,6 +1360,42 @@ IV = 9a785634120000000000000000000000
 Plaintext = 000102030405060708090a0b0c0d0e0f10111213
 Ciphertext = 9d84c813f719aa2c7be3f66171c7c5c2edbf9dac
 
+Cipher = aes-128-xts
```
Comment (levitte) [inline test/recipes/30-test_evp_data/evpciph.txt:1363]: Are these test vectors also from IEEE Std 1619-2007 (see the Title on line 1229)?  If not, a new title line is required

**Comment by levitte** [comment]
@xkqian, sorry for having left you hanging...  I'm not good enough to evaluate the contents of the new assembler, other than saying that the form look good.  Other than that, I can only take it at face value...  and that's ok.

**Comment by levitte** [comment]
That being said, you will unfortunately need to rebase...

**Comment by xkqian** [comment]
> That being said, you will unfortunately need to rebase...

No problem, I will rebase and add more test datas for other micro-architectures.

File: test/recipes/30-test_evp_data/evpciph.txt (hunk 11 above)
Comment (xkqian) [reply test/recipes/30-test_evp_data/evpciph.txt:1363]: These added test vectors are not from the IEEE Std 1619-2007, maybe I should add one new title. Thanks.  Is it fine if I add title like : "AES XTS Non standard test vectors - generated from reference implementation" ?

Comment (levitte) [reply test/recipes/30-test_evp_data/evpciph.txt:1363]: That's a good title, yes 🙂

**Comment by xkqian** [comment]
One of the travis CI failed, but seems it faild on the AMD64 on this case: "30-test_evp_fetch_prov.t ", and had nothing to do with this patch,  so please help review ignore the failure, thanks.

**Review by paulidale — APPROVED** [review]
I also don't have a good knowledge of the assembler.  A symmetric algorithm that passes the unit tests will be at least _mostly_ good.

**Comment by ardbiesheuvel** [comment]
Could you explain the significance of the 3-way interleaved code path? Does it ever execute more than a single time for each invocation of the XTS en/decrypt routine? Because I'd assume that it only gets called if the non-multiple-of-5 remainder of the number of blocks is 3 or 4, and we can probably do without it.

**Comment by xkqian** [comment]
> Could you explain the significance of the 3-way interleaved code path? Does it ever execute more than a single time for each invocation of the XTS en/decrypt routine? Because I'd assume that it only gets called if the non-multiple-of-5 remainder of the number of blocks is 3 or 4, and we can probably do without it.
Here, the 3-way interleaved code is for tail processing, sure, it gets called only once every en/decrypt routine when the "remainder of (n mod 5)" equals to 3.  Besides, we also have 2-way interlaved code if remainder equals to 2 or 1.  Our design standard is try to increase the performance gain and avoid the regression even in case of small sizes. If we drop the 3-way interleaved code path,  the performance will  have regession when remainder equals to 3 compared with current implementation,  in some special case such as data size= (8*16) bytes, the drop even reachs 30%.

**Comment by ardbiesheuvel** [comment]
I am not convinced that saving up to ~30 cycles (assuming ~1 cpb performance) justifies the increased code size and complexity. Are you aware of any real use cases where XTS encryption of 128 byte blocks is on a hot path, and where this 30 cycle difference is measurable in the overall performance?

Benchmarks are important, but on a real system, larger code means large I-cache footprint as well, which penalizes other tasks competing for the CPU.

**Comment by paulidale** [comment]
The 3-way folded path does seem excessive.  The benefit for practical XTS block sizes seem like they'd be very small -- low single digit percentage for fairly small block going down to nothing for larger.  XTS is designed for file system encryption -- I'd expect 256 or 512 byte blocks as a minimum and likely larger.

**Comment by xkqian** [comment]
> The 3-way folded path does seem excessive. The benefit for practical XTS block sizes seem like they'd be very small -- low single digit percentage for fairly small block going down to nothing for larger. XTS is designed for file system encryption -- I'd expect 256 or 512 byte blocks as a minimum and likely larger.


(256/16) mod 5 = 1, and (512/16) mod 5 =2, both of them are't affected by the 3-way folded path, it seems reasonable to drop the 3-way.

**Comment by xkqian** [comment]
> I am not convinced that saving up to ~30 cycles (assuming ~1 cpb performance) justifies the increased code size and complexity. Are you aware of any real use cases where XTS encryption of 128 byte blocks is on a hot path, and where this 30 cycle difference is measurable in the overall performance?
> 
> Benchmarks are important, but on a real system, larger code means large I-cache footprint as well, which penalizes other tasks competing for the CPU.

I didn't find that 128 bytes is in the hot path of enc/decryption, what we did is utilizing the max potent of the pipelines to optimize the algorithms.  Cause the 3-way isn't in the not path, anyway, I will drop it in the later version.

**Comment by xkqian** [comment]
I am not sure what changes happened in the upstream, with latest code, plush my PR, all of the aes-xts test cases will fail. I will fix it in the same commit with the 3-way fold code path drop.

**Comment by xkqian** [comment]
> I am not sure what changes happened in the upstream, with latest code, plush my PR, all of the aes-xts test cases will fail. I will fix it in the same commit with the 3-way fold code path drop.

Seems it's the commirt 4b51903d8681c7fd429c566548529d5753e24f47 which cause the regression. It's wierd that the code

>  {
>         XTS_SET_KEY_FN(AES_set_encrypt_key, AES_set_decrypt_key,
>                        AES_encrypt, AES_decrypt, stream_enc, stream_dec);
>     } 

will be called whether it's aarch64 or x86. We think it can be wrapped in x86 use only just as the origianl version did.
The commit as below:

> ommit 4b51903d8681c7fd429c566548529d5753e24f47 (HEAD)
> Author: Shane Lontis <shane.lontis@oracle.com>
> Date:   Wed Sep 16 11:07:02 2020 +1000
> 
>     Fix AES_XTS on x86-64 platforms with BSAES and VPAES support.
> 
>     Fixes #11622
>     Fixes #12378
> 
>     Due to a missing else it was setting up the stream for BSAES and then using this incorrect stream with VPAES.
>     The correct behaviour is not to use VPAES at all in this case.
>     Also note that the original code in e_aes could set up VPAES and then would overwrite it with the generic implementation.
>     On a machine that supported both BSAES and VPAES the code was changed locally to force it to run both cases to verify
>     both paths produce the correct known answers.
> 
>     Debugged using mageia 7.1, but is also highly likely to fix FreeBSD also.
> 
>     Reviewed-by: Tomas Mraz <tmraz@fedoraproject.org>
>     (Merged from https://github.com/openssl/openssl/pull/12887)

**Comment by kaduk** [comment]
> I am not sure what changes happened in the upstream, with latest code, plush my PR, all of the aes-xts test cases will fail. I will fix it in the same commit with the 3-way fold code path drop.

I would suggest to use separate commits, even if you push them at the same time.  Separate commits makes it easier to see the different changes in isolation.

**Comment by xkqian** [comment]
> > I am not sure what changes happened in the upstream, with latest code, plush my PR, all of the aes-xts test cases will fail. I will fix it in the same commit with the 3-way fold code path drop.
> 
> I would suggest to use separate commits, even if you push them at the same time. Separate commits makes it easier to see the different changes in isolation.

Sure, I will commit them in different patches. Firstly prepare one new patch to fix the potential issues which maybe introduced by upstream, and then back to this optimization patch.

**Comment by xkqian** [comment]
Ping... Will someone kindly jump into the review?

**Comment by t8m** [comment]
Please rebase to fix the conflict.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@paulidale  does your review still hold?

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Thanks for the submission.

**Pull request closed without merging**